File: 1299-k-concatenation-maximum-sum/k-concatenation-maximum-sum.py
class Solution:
    def kConcatenationMaxSum(self, arr: List[int], k: int) -> int:
        sum_arr = sum(arr)
        MOD = 10**9 + 7

        def get_kadane(data):
            curr_max = res = 0 
            for x in data:
                curr_max = max(x, curr_max + x)
                res = max(res, curr_max)
            return res

        if k == 1:
            return get_kadane(arr) % MOD

        if sum_arr <= 0:
            return get_kadane(arr + arr) % MOD
        
        return (get_kadane(arr + arr) + (k - 2) * sum_arr) % MOD

        


        
        

File: 1299-k-concatenation-maximum-sum/k-concatenation-maximum-sum.cpp
class Solution {
public:
    int kadane(vector<int>& arr)
    {
        long long currMax = 0;
        long long ansMax = 0;

        for(int i=0 ; i < arr.size() ;i++)
        {
            currMax = max(currMax + arr[i] , (long long)arr[i]);
            ansMax = max(ansMax,currMax);
        }

        return ansMax;
    }

    int kConcatenationMaxSum(vector<int>& arr, int k) 
    {
        long long sum = accumulate(arr.begin(),arr.end(),0);
        long long MOD = 1e9 + 7;

        if ( k == 1)
        {
            return kadane(arr) % MOD;
        }

        vector<int> newArr = arr;
        newArr.insert(newArr.end(),arr.begin(),arr.end());

        long long maxDouble = kadane(newArr);

        if (sum > 0)
        {
            return ((long long) (k-2) * sum  + maxDouble) % MOD;
        }

        return (maxDouble) % MOD;
    }
};